cmake/Re2cBuildType.cmake

```diff
@@ -1,14 +1,85 @@
+    set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "${_allowed_build_types}")
+
+    # - The default value for CMAKE_BUILD_TYPE is an empty string
+    # - User can set CMAKE_BUILD_TYPE to any value at the cmake command line
+    #
+    # Therefore, we check both cases, and make sure that we are dealing with
+    # a known build type (if any).
+    #
+    # For more see discussion at https://github.com/skvadrik/re2c/pull/316
+    if(CMAKE_BUILD_TYPE AND NOT CMAKE_BUILD_TYPE IN_LIST _allowed_build_types)
+        message(FATAL_ERROR "Invalid build type: ${CMAKE_BUILD_TYPE}")
+    endif()
+
+    unset(_allowed_build_types)
+elseif(NOT MULTICONFIG_DONE) # Xcode, Visual Studio, ...
+    set(MULTICONFIG_DONE TRUE)
+    if (NOT CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
+        list(APPEND CMAKE_CONFIGURATION_TYPES Asan Ubsan)
+
+        # This is needed because user can set CMAKE_CONFIGURATION_TYPES
+        # at the cmake command line
+        list(REMOVE_DUPLICATES CMAKE_CONFIGURATION_TYPES)
+
+        set(CMAKE_CONFIGURATION_TYPES "${CMAKE_CONFIGURATION_TYPES}" CACHE STRING
+            "Semicolon separated list of supported configuration types." FORCE)
+    endif()
+endif()
+
+# Notes for Windows with Visual Studio:
 #
-# Therefore, we check both cases, and make sure that we are dealing with
-# a known build type (if any).
+# - Address Sanitizer (Asan) currently is under experimental stage
+# - Undefined Behavior Sanitizer (Ubsan) is not supported at this time
 #
-# For more see discussion at https://github.com/skvadrik/re2c/pull/316
-if(CMAKE_BUILD_TYPE AND NOT CMAKE_BUILD_TYPE IN_LIST _allowed_build_types)
-    message(FATAL_ERROR "Invalid build type: ${CMAKE_BUILD_TYPE}")
+# Therefore, we have disabled support for these build types now, but
+# this support may be enabled in the future.
+if (NOT CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
+    # Setup Asan flags
+    set(CMAKE_C_FLAGS_ASAN
+        "${CMAKE_C_FLAGS_DEBUG} -fsanitize=address -fno-omit-frame-pointer" CACHE STRING
+        "Flags used by the C compiler for Asan build type or configuration." FORCE)
+
+    set(CMAKE_CXX_FLAGS_ASAN
+        "${CMAKE_CXX_FLAGS_DEBUG} -fsanitize=address -fno-omit-frame-pointer" CACHE STRING
+        "Flags used by the C++ compiler for Asan build type or configuration." FORCE)
+
+    set(CMAKE_EXE_LINKER_FLAGS_ASAN
+        "${CMAKE_EXE_LINKER_FLAGS_DEBUG} -fsanitize=address -fno-omit-frame-pointer" CACHE STRING
+        "Linker flags to be used to create executables for Asan build type." FORCE)
+
+    set(CMAKE_SHARED_LINKER_FLAGS_ASAN
+        "${CMAKE_SHARED_LINKER_FLAGS_DEBUG} -fsanitize=address -fno-omit-frame-pointer" CACHE STRING
+        "Linker flags to be used to create shared libraries for Asan build type." FORCE)
+
+    set(CMAKE_STATIC_LINKER_FLAGS_ASAN
+        "${CMAKE_STATIC_LINKER_FLAGS_DEBUG} -fsanitize=address -fno-omit-frame-pointer" CACHE STRING
+        "Linker flags to be used to create static libraries for Asan build type." FORCE)
```

Comment:
Is `-fno-omit-frame-pointer` needed for linker?

Reply:
I've updated this part - added notes and removed linker flag. Please take a look.